src/codeblockediting.js

```diff
@@ -0,0 +1,409 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module code-block/codeblockediting
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import ShiftEnter from '@ckeditor/ckeditor5-enter/src/shiftenter';
+import CodeBlockCommand from './codeblockcommand';
+import IndentCodeBlockCommand from './indentcodeblockcommand';
+import {
+	getLocalizedLanguageDefinitions,
+	getLeadingWhiteSpaces,
+	rawSnippetTextToModelDocumentFragment
+} from './utils';
+import {
+	modelToViewCodeBlockInsertion,
+	modelToDataViewSoftBreakInsertion,
+	dataViewToModelCodeBlockInsertion
+} from './converters';
+
+const DEFAULT_ELEMENT = 'paragraph';
+
+/**
+ * The editing part of the code block feature.
+ *
+ * Introduces the `'codeBlock'` command and the `'codeBlock'` model element.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class CodeBlockEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'CodeBlockEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ ShiftEnter ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		editor.config.define( 'codeBlock', {
+			languages: [
+				{ class: 'plaintext', label: 'Plain text' },
+				{ class: 'c', label: 'C' },
+				{ class: 'cs', label: 'C#' },
+				{ class: 'cpp', label: 'C++' },
+				{ class: 'css', label: 'CSS' },
+				{ class: 'diff', label: 'Diff' },
+				{ class: 'xml', label: 'HTML/XML' },
+				{ class: 'java', label: 'Java' },
+				{ class: 'javascript', label: 'JavaScript' },
+				{ class: 'php', label: 'PHP' },
+				{ class: 'python', label: 'Python' },
+				{ class: 'ruby', label: 'Ruby' },
+				{ class: 'typescript', label: 'TypeScript' },
+			],
+
+			// A single tab.
+			indentSequence: '	'
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+		const model = editor.model;
+
+		const localizedLanguageDefinitions = getLocalizedLanguageDefinitions( editor );
+		const languageClasses = localizedLanguageDefinitions.map( def => def.class );
+		const languageLabels = Object.assign( {}, ...localizedLanguageDefinitions.map( def => ( { [ def.class ]: def.label } ) ) );
+
+		// The main command.
+		editor.commands.add( 'codeBlock', new CodeBlockCommand( editor ) );
```

Comment:
Maybe `insertCodeBlock`?

Reply:
For the command name, I think we might stay with `codeBlock` as `insert...` would suggest inserting something into the content (AFAIR there's `insertImage`, `insertTable).

Here `CodeBlock` behaves similarly to other features that transform/enhances some text. You will create a code block based on a selection (similarly to block quote).

In other words there's no "inserting" of new or created content.